### PR TITLE
Extract and collect error codes from ManagedResources

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
This PR makes the `WaitUntilManagedResourcesDeleted` to detect and extract error codes from the `ManagedResources` that is stuck in deletion.

**Release note**:

```feature operator
The error code detection has been enhanced for `ManagedResource` objects that are stuck when a shoot is deleted. This enables Gardener to assign the corresponding error code(s) to the shoot object.
```

### CODE REVIEW

File: pkg/apis/core/v1beta1/helper/errors.go [hunk 1]
```diff
@@ -146,6 +146,62 @@ func ExtractErrorCodes(err error) []gardencorev1beta1.ErrorCode {
 	return codes
 }
 
+var _ error = (*MultiErrorWithCodes)(nil)
+
+// MultiErrorWithCodes is a struct that contains multiple errors and `ErrorCodes`s.
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/helper/errors.go]:
```suggestion
// MultiErrorWithCodes is a struct that contains multiple errors and ErrorCodes.
```

File: pkg/gardenlet/controller/shoot/shoot_control_delete.go [hunk 2]
```diff
@@ -388,7 +388,7 @@ func (r *shootReconciler) runDeleteShootFlow(ctx context.Context, o *operation.O
 		})
 		cleanShootNamespaces = g.Add(flow.Task{
 			Name:         "Cleaning shoot namespaces",
-			Fn:           flow.TaskFn(botanist.CleanShootNamespaces).Timeout(10 * time.Minute).DoIf(cleanupShootResources),
+			Fn:           flow.TaskFn(botanist. CleanShootNamespaces).Timeout(10 * time.Minute).DoIf(cleanupShootResources),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot_control_delete.go]: Revert this

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by timuthy** [comment]
Thanks for your review @rfranzke. PTAL.

File: pkg/apis/core/v1beta1/helper/errors.go [hunk 3]
```diff
@@ -148,7 +148,7 @@ func ExtractErrorCodes(err error) []gardencorev1beta1.ErrorCode {
 
 var _ error = (*MultiErrorWithCodes)(nil)
 
-// MultiErrorWithCodes is a struct that contains multiple errors and `ErrorCodes`s.
+// MultiErrorWithCodes is a struct that contains multiple errors and ErrorCodess.
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/helper/errors.go]:
```suggestion
// MultiErrorWithCodes is a struct that contains multiple errors and ErrorCodes.
```

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**